src/equations/compressible_euler_2d.jl

```diff
@@ -1957,6 +1957,28 @@ end
     return rho
 end
 
+@inline function velocity(u, equations::CompressibleEulerEquations2D)
+    rho = u[1]
+    v1 = u[2] / rho
+    v2 = u[3] / rho
+    return SVector(v1, v2)
+end
+
+@inline function velocity(u, orientation::Int, equations::CompressibleEulerEquations2D)
+    rho = u[1]
+    v = u[orientation + 1] / rho
+    return v
+end
+
+@inline function velocity(u, normal_direction::AbstractVector,
+                          equations::CompressibleEulerEquations2D)
+    rho = u[1]
+    v1 = u[2] / rho
+    v2 = u[3] / rho
+    v = v1 * normal_direction[1] + v2 * normal_direction[2]
+    return v
+end
```

Comment:
Why do we need three versions of this functionality? Wouldn't it be sufficient to use the one that only dispatches on the equation set and returns the `SVector` of Cartesian fluxes. Then computing a particular normal direction, using orientation or a vector, could be done within the function that calls `velocity`.

Reply:
Thank you for the observation!

I updated the PR to have a general function for the velocity in `normal_direction`.

However, having different versions for the `orientation` is slightly cheaper, as the general `velocity` function will compute velocity in all directions even though we need only one. Thus, there is a minor readability-performance trade-off and I would like your thoughts on it.

Reply:
I can see a reason for this. Checking a simple example, the compiler seems to be smart enough to optimize away the case of constant `orientation`s but not with variable ones, e.g., 

```julia
julia> using StaticArrays

julia> @inline function velocity(u)
           rho = u[1]
           v1 = u[2] / rho
           v2 = u[3] / rho
           return SVector(v1, v2)
       end
velocity (generic function with 1 method)

julia> @inline function velocity(u, orientation)
           rho = u[1]
           v = u[1 + orientation] / rho
           return v
       end
velocity (generic function with 2 methods)

julia> foo1(u) = velocity(u)[1]
foo1 (generic function with 1 method)

julia> foo2(u) = velocity(u, 1)
foo2 (generic function with 1 method)

julia> u = SVector(1.0, 0.1, -0.2, 1.5)
4-element SVector{4, Float64} with indices SOneTo(4):
  1.0
  0.1
 -0.2
  1.5

julia> @code_llvm debuginfo=:none foo1(u)
define double @julia_foo1_188([1 x [4 x double]]* nocapture noundef nonnull readonly align 8 dereferenceable(32) %0) #0 {
top:
  %1 = getelementptr inbounds [1 x [4 x double]], [1 x [4 x double]]* %0, i64 0, i64 0, i64 0
  %2 = getelementptr inbounds [1 x [4 x double]], [1 x [4 x double]]* %0, i64 0, i64 0, i64 1
  %unbox = load double, double* %2, align 8
  %unbox1 = load double, double* %1, align 8
  %3 = fdiv double %unbox, %unbox1
  ret double %3
}

julia> @code_llvm debuginfo=:none foo2(u)
define double @julia_foo2_201([1 x [4 x double]]* nocapture noundef nonnull readonly align 8 dereferenceable(32) %0) #0 {
top:
  %1 = getelementptr inbounds [1 x [4 x double]], [1 x [4 x double]]* %0, i64 0, i64 0, i64 0
  %2 = getelementptr inbounds [1 x [4 x double]], [1 x [4 x double]]* %0, i64 0, i64 0, i64 1
  %unbox = load double, double* %2, align 8
  %unbox1 = load double, double* %1, align 8
  %3 = fdiv double %unbox, %unbox1
  ret double %3
}

julia> bar1(u, i) = velocity(u)[i]
bar1 (generic function with 1 method)

julia> bar2(u, i) = velocity(u, i)
bar2 (generic function with 1 method)

julia> @code_llvm debuginfo=:none bar1(u, 1)
define double @julia_bar1_215([1 x [4 x double]]* nocapture noundef nonnull readonly align 8 dereferenceable(32) %0, i64 signext %1) #0 {
top:
  %newstruct = alloca <2 x double>, align 16
  %2 = getelementptr inbounds [1 x [4 x double]], [1 x [4 x double]]* %0, i64 0, i64 0, i64 0
  %3 = getelementptr inbounds [1 x [4 x double]], [1 x [4 x double]]* %0, i64 0, i64 0, i64 1
  %unbox1 = load double, double* %2, align 8
  %4 = bitcast double* %3 to <2 x double>*
  %5 = load <2 x double>, <2 x double>* %4, align 8
  %6 = insertelement <2 x double> poison, double %unbox1, i64 0
  %7 = shufflevector <2 x double> %6, <2 x double> poison, <2 x i32> zeroinitializer
  %8 = fdiv <2 x double> %5, %7
  store <2 x double> %8, <2 x double>* %newstruct, align 16
  %9 = add i64 %1, -1
  %boundscheck = icmp ult i64 %9, 2
  br i1 %boundscheck, label %pass, label %fail

fail:                                             ; preds = %top
  %10 = bitcast <2 x double>* %newstruct to i8*
  call void @ijl_bounds_error_unboxed_int(i8* %10, {}* nonnull inttoptr (i64 4803813760 to {}*), i64 %1)
  unreachable

pass:                                             ; preds = %top
  %11 = getelementptr inbounds <2 x double>, <2 x double>* %newstruct, i64 0, i64 %9
  %unbox4 = load double, double* %11, align 8
  ret double %unbox4
}

julia> @code_llvm debuginfo=:none bar2(u, 1)
define double @julia_bar2_217([1 x [4 x double]]* nocapture noundef nonnull readonly align 8 dereferenceable(32) %0, i64 signext %1) #0 {
top:
  %boundscheck = icmp ult i64 %1, 4
  br i1 %boundscheck, label %pass, label %fail

fail:                                             ; preds = %top
  %2 = add i64 %1, 1
  %3 = bitcast [1 x [4 x double]]* %0 to i8*
  call void @ijl_bounds_error_unboxed_int(i8* nonnull %3, {}* nonnull inttoptr (i64 4766129648 to {}*), i64 %2)
  unreachable

pass:                                             ; preds = %top
  %4 = getelementptr inbounds [1 x [4 x double]], [1 x [4 x double]]* %0, i64 0, i64 0, i64 0
  %5 = getelementptr inbounds [1 x [4 x double]], [1 x [4 x double]]* %0, i64 0, i64 0, i64 %1
  %unbox = load double, double* %5, align 8
  %unbox1 = load double, double* %4, align 8
  %6 = fdiv double %unbox, %unbox1
  ret double %6
}
```

Reply:
What do you think, @andrewwinters5000?

Reply:
That looks reasonable. I like the new strategy of having generic versions for the `normal_direction` variant.